### PR TITLE
[WIP] remove expensive api from InternalRow

### PR DESCRIPTION
**Pull request by davies**
This is a follow up PR for #6792, to use narrow down the APIs in InternalRow (remove those expensive ones).

### CODE REVIEW

**Comment by davies** [comment]
cc @liancheng @rxin @JoshRosen

**Comment by cloud-fan** [comment]
hi @davies , I met a bug at https://github.com/apache/spark/pull/6825 that we didn't handle `BinaryType(byte[])` correctly at `equals` and `hashCode`. We should use `java.util.Arrays.equals` and `java.util.Arrays.hashCode`. Could you fix it here?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/InternalRow.scala [hunk 1]
```diff
@@ -19,39 +19,193 @@ package org.apache.spark.sql.catalyst
...
+ *   MapType -> scala.collection.Map
+ *   StructType -> InternalRow
  */
-abstract class InternalRow extends Row {
-  // A default implementation to change the return type
-  override def copy(): InternalRow = {this}
-}
+abstract class InternalRow extends Serializable {
+  /** Number of elements in the Row. */
+  def size: Int = length
```
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/InternalRow.scala]: do we need size? maybe just keep one of length or size.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/InternalRow.scala [hunk 2]
```diff
@@ -19,39 +19,193 @@ package org.apache.spark.sql.catalyst
...
-  def unapplySeq(row: InternalRow): Some[Seq[Any]] = Some(row.toSeq)
+  /** Number of elements in the Row. */
+  def length: Int
 
   /**
-   * This method can be used to construct a [[Row]] with the given values.
+   * Schema for the row.
    */
-  def apply(values: Any*): InternalRow = new GenericRow(values.toArray)
+  def schema: StructType = null
```
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/InternalRow.scala]: do we need this here?

File: sql/core/src/main/scala/org/apache/spark/sql/sources/commands.scala [hunk 3]
```diff
@@ -143,14 +143,15 @@ private[sql] case class InsertIntoHadoopFsRelation(
...
-          r: InternalRow => r.asInstanceOf[Row]
-        }
+        val converter = CatalystTypeConverters.createToScalaConverter(dataSchema)
+          .asInstanceOf[InternalRow => Row]
         while (iterator.hasNext) {
-          val row = converter(iterator.next())
-          writerContainer.outputWriterForRow(row).write(row)
+          val row = iterator.next()
+          val converted = converter(row)
+          writerContainer.outputWriterForRow(converted) match {
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/sources/commands.scala]: is this expensive?

File: sql/core/src/main/scala/org/apache/spark/sql/sources/interfaces.scala [hunk 4]
```diff
@@ -342,6 +342,16 @@ abstract class OutputWriter {
 }
 
 /**
+ * internal only
+ */
+abstract class InternalOutputWriter extends OutputWriter {
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/sources/interfaces.scala]: who implements this?

**Comment by rxin** [comment]
@davies can you update the pr description to include what you actually removed?

**Comment by davies** [comment]
@cloud-fan it will be fixed by https://github.com/apache/spark/pull/6876

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/InternalRow.scala (hunk 1 above)
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/InternalRow.scala]: agree. It's internal and we don't need to follow scala convention to provide `size`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/InternalRow.scala [hunk 5]
```diff
@@ -19,39 +19,193 @@ package org.apache.spark.sql.catalyst
...
-    new GenericRow(rows.flatMap(_.toSeq).toArray)
+  def getDouble(i: Int): Double
+
+  /**
+   * Returns the value at position i as a String object.
+   *
+   * @throws ClassCastException when data type does not match.
+   * @throws NullPointerException when value is null.
+   */
+  def getString(i: Int): String = getAs[UTF8String](i).toString
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/InternalRow.scala]: As it's internal row, should we just return `UTF8String`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/InternalRow.scala [hunk 6]
```diff
@@ -19,39 +19,193 @@ package org.apache.spark.sql.catalyst
...
+
+  /**
+   * Returns the value at position i.
+   *
+   * @throws ClassCastException when data type does not match.
+   */
+  def getAs[T](i: Int): T = apply(i).asInstanceOf[T]
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/InternalRow.scala]: will this throw `NullPointerException`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/Row.scala [hunk 7]
```diff
@@ -153,7 +152,7 @@ trait Row extends Serializable {
    *   StructType -> org.apache.spark.sql.Row
    * }}}
    */
-  def apply(i: Int): Any
+  override def apply(i: Int): Any
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/Row.scala]: why we override it without implement it?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/InternalRow.scala [hunk 8]
```diff
@@ -19,39 +19,193 @@ package org.apache.spark.sql.catalyst
...
+    val len = length
+    var i = 0
+    while (i < len) {
+      if (isNullAt(i)) { return true }
+      i += 1
+    }
+    false
+  }
+
+  override def equals(that: Any): Boolean = that match {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/InternalRow.scala]: Is it OK to remove the hashCode implementation?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/InternalRow.scala (hunk 5 above)
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/InternalRow.scala]: Maybe we can only define `getUTF8String` at `InternalRow` and define `getString` at user-facing `Row`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/SpecificMutableRow.scala [hunk 9]
```diff
@@ -191,7 +192,7 @@ final class MutableAny extends MutableValue {
  * based on the dataTypes of each column.  The intent is to decrease garbage when modifying the
  * values of primitive columns.
  */
-final class SpecificMutableRow(val values: Array[MutableValue]) extends MutableRow {
+final class SpecificMutableRow(val values: Array[MutableValue]) extends Row with MutableRow {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/SpecificMutableRow.scala]: why not `extends InternalRow with MutableRow`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/rows.scala [hunk 10]
```diff
@@ -54,7 +56,6 @@ object EmptyRow extends InternalRow {
   override def getBoolean(i: Int): Boolean = throw new UnsupportedOperationException
   override def getShort(i: Int): Short = throw new UnsupportedOperationException
   override def getByte(i: Int): Byte = throw new UnsupportedOperationException
-  override def getString(i: Int): String = throw new UnsupportedOperationException
   override def getAs[T](i: Int): T = throw new UnsupportedOperationException
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/rows.scala]: `getAs` wil call `apply`, so we can also remove it?

**Comment by davies** [comment]
Close this, in favor of #7003

**Pull request closed without merging**